### PR TITLE
gql to pydantic

### PR DESCRIPTION
**Pull request by doctrino**
#### Added

* Support for converting from `.graphql` to `pydantic`. This means you can now write your types in a `.graphql` file
  and automatically generate the `pytandic` counterpart.
* CLI for doing the conversion between GraphQL and `pydantic`. The following two commands
  * `pygen togql` converts a pydantic schema to `.graphql` and creates the `client.py` which is the SDK to interact
    with the data model through Python.
  * `pygen topython` converts a `.graphql` to `pytandic` classes. This command also creates the `client.py`, i.e., the
    SDK to interact with the data model through Pytho
* Documentation of motivation and usage of package.

#### Changed

* CLI command going from `gqlpygen` to `pygen`. This is for ease of use.

#### Deviation from Original Plan

This PR ended up deviating from the original plan of investigating the open source alternatives. The reason is that this is time consuming and writing the current functionality was a low hanging fruit. After this PR is merged, there is still no GraphQL
support for querying your data model, but it becomes easier to populate it with data and discover what is available. 

### Original Plan
The goal is to create code to generate a python SDK from `.graphql` input. Trying to initially go a simple scope as follows:

 * take the input in the from the `examples/cinematography_domain/schema.graphql` and generate the `client.py` and `schema.py` files.

Looking for feedback. 

Current plan is to investigate existing packages, to see if a quick wrapping can do the trick. Most of these are primarily for backend implementation, but at least some of them have a codegen feature. 

* [strawberry](https://strawberry.rocks/docs/codegen/query-codegen)
    * Seems to require schema built in strawberry.
* [ariadne](https://ariadnegraphql.org/)
    * Input schema + queries -> pydantic classes. Requires mutations. 
* [tartiflette](https://tartiflette.io/)
* [graphene](https://graphene-python.org/)
* [gql](https://gql.readthedocs.io/en/stable/intro.html)

## Checklist:
- [x] Tests added/updated.
- [x] Documentation updated. Documentation is generated from docstrings - these must be updated according to your change.
  If a new method has been added it should be referenced in [cognite.rst](https://github.com/cognitedata/cognite-sdk-python/blob/master/docs/source/cognite.rst) in order to generate docs based on its docstring.
- [x] Changelog updated in [CHANGELOG.md](https://github.com/cognitedata/cognite-sdk-python/blob/master/CHANGELOG.md).
- [x] Version bumped. If triggering a new release is desired, bump the version number in [_version.py](https://github.com/cognitedata/cognite-sdk-python/blob/master/cognite/client/_version.py) and [pyproject.toml](https://github.com/cognitedata/cognite-sdk-python/blob/master/pyproject.toml) per [semantic versioning](https://semver.org/).

### CODE REVIEW

File: cognite/dm_clients/domain_modeling/schema.py [hunk 1]
```diff
@@ -17,6 +17,10 @@
 
 
 logger = logging.getLogger(__name__)
+AUTO_GENERATED_COMMENT = """# THIS FILE IS AUTO-GENERATED!
+# Use `dm_clients schema render` to update it, see `dm_clients --help` for more information.
+
+"""
```
Comment (frnhr) [inline cognite/dm_clients/domain_modeling/schema.py:23]: This comment is now duplicated when generating the schema using the cli tool `dm_clients schema render`. We should redo the scripts in pure Python, but for now maybe we can just remove the comment from `dm_clients/bin/schema_render.sh`?

**Review by frnhr — CHANGES_REQUESTED** [review]

File: examples/cinematography_domain/client.py [hunk 2]
```diff
@@ -17,8 +17,8 @@ class CineClient(DomainClient):
     `movie`) are create in the base class, and here we have only added type annotations so that IntelliSense can work.
     """
 
-    person: DomainModelAPI[Person]
     movie: DomainModelAPI[Movie]
+    person: DomainModelAPI[Person]
```
Comment (frnhr) [inline examples/cinematography_domain/client.py:21]: Was this change required? Seems to work will in both cases.

File: README.md [hunk 3]
```diff
@@ -4,35 +4,82 @@ Cognite GraphQL Python Generator
...
+```
+
+This will create a `schema.py` and a `client.py` file in the directory you are running the command.
+
+To generate from `schema.py` use the following command
 
-Type the following command for help.
 ```bash
-cogpygen --help
+pygen togql 'PATH_TO_FILE'
```
Comment (frnhr) [inline README.md:55]: Just thinking (outside of this PR' scope), `pygen` is not the greatest name for a CLI tool, and neither is `dm_clients` (as in `dm_clients schema render`). Any ideas for a common CLI tool name? For allthethings, like:
  * `tool_name togql PATH_TO_FILE`
  * `tool_name topy ...`
  * `tool_name to_cdf ...` or `tool_name publish`  or `tool_name upload` ?

File: README.md [hunk 4]
```diff
@@ -4,35 +4,82 @@ Cognite GraphQL Python Generator
...
+1. Instantiate a new schema with the line, `myschema: Schema[DomainModel] = Schema()`
+2. Register all you Types with `@myschema.register_type`
+3. Close the schema with `myschema.close()`
+
+To get a concrete example is available in [examples/cinematography_domain](https://github.com/cognitedata/cognite-gql-pygen/blob/main/examples/cinematography_domain),
+it consists of four files.
+
+* `schema.graphql` The schema defined in GraphQL language.
+* `schema.py` The schema defined in `pydantic` classes.
+* `client.py` Which sets ut the client for the data model.
```
Comment (frnhr) [inline README.md:72]:
```suggestion
* `client.py` Which sets up the client for the data model.
```
typo

File: cognite/dm_clients/domain_modeling/schema.py (hunk 1 above)
Comment (doctrino) [reply cognite/dm_clients/domain_modeling/schema.py:23]: done.

File: examples/cinematography_domain/client.py (hunk 2 above)
Comment (doctrino) [reply examples/cinematography_domain/client.py:21]: Just for testing purposes., I sort the order based on name to get consistent results.

File: README.md (hunk 3 above)
Comment (doctrino) [reply README.md:55]: I am leaning towards leaving some of this functionalities to the already existing CLI, cdf. Then, this name, or coming up with a better name will be easier.

Comment (doctrino) [reply README.md:55]: I guess just `dm` could work. wdyt?

**Review by frnhr — APPROVED** [review]
LGTM

File: README.md (hunk 3 above)
Comment (doctrino) [reply README.md]: I decided `dm` is a better name, so changed it. `dm togql`, `dm topython`, `dm publish`, dm upload all works better.
